modalc-lib/modal-vector.rkt

```diff
@@ -0,0 +1,118 @@
+#lang racket
+
+(require modalc
+         syntax/parse/define)
+
+(define (modal-vector/c should-apply-ctc? . inner-ctc)
+  (define inner-ctc-proj
+    (map contract-late-neg-projection inner-ctc))
+  (make-contract
+   #:name `(modal-vector/c ,(map contract-name inner-ctc-proj))
+   #:late-neg-projection
+   (λ (blame)
+     (define inner-ctc-proj/blame (map (λ (p) (p (blame-swap blame))) inner-ctc-proj))
+     (λ (val neg-party)
+       (impersonate-vector val
+                           (λ (vec index value)
+                             (cond [(should-apply-ctc? (list index value))
+                                    (printf "ref call: ~s\n" (list index value))
+                                    (for/list ([arg (in-vector vec)]
+                                               [proj (in-list inner-ctc-proj/blame)])
+                                      (proj arg neg-party))
+                                    value]
+                                   [else (printf "no ref ~s\n" (list index value))
+                                         value]))
+                           (λ (vec index value)
+                             (cond [(should-apply-ctc? (list index value))
+                                    (printf "set call: ~s\n" (list index value))
+                                    (for/list ([arg (in-vector vec)]
+                                               [proj (in-list inner-ctc-proj/blame)])
+                                      (proj arg neg-party))
+                                    (vector-set! vec index value)
+                                    value]
+                                   [else (printf "no set ~s\n" (list index value))
+                                         value])))))))
+
+(define-simple-macro (modal-vector/c* mode
+                                 {~and inner-ctc {~not _:keyword}} ...
+                                 {~seq kw:keyword ctc} ...)
+  (make-modal-vector/c* mode
+                   (list inner-ctc ...)
+                   (list (list 'kw ctc) ...)))
+
+(define (make-modal-vector/c* should-apply-ctc? inner-ctc kwds)
```

Comment:
Why not just have `modal-vector/c*` be a function that accepts regular keyword arguments? No need for this macro then. Here's the kind of pattern I think is idiomatic.
```
(define (modal-vector/c* #:ref [maybe-ref-mode #f]
                         #:set [maybe-set-mode #f]
                         #:ref/set [maybe-both-mode #f]
                         . inner-ctcs)
  (unless (or maybe-ref-mode
              maybe-set-mode
              maybe-both-mode)
    (raise-user-error 'modal-vector/c* "no mode supplied"))
  (define should-apply-ctc?/ref (or maybe-ref-mode
                                    maybe-both-mode
                                    mode:never))
  (define should-apply-ctc?/set (or maybe-set-mode
                                    maybe-both-mode
                                    mode:never))
  ;; ... use /ref and /set below
```